### DISCUSSION TITLE
Want to use getServerProps instead of getInitialProps, but how to implement server side redirect?

### DISCUSSION THREAD

**Discussion by oran1248** [vercel/next.js#13635] [Help]
Hi,

Today, I'm performing authentiocation checks of a `page` in  `getInitialProps`. if the user is not authenticated, i'm redirecting him to another page. 
I read the documentation of `getServerProps` and saw the recommendation to use it and not `getInitialProps`. 
But, how can I implement server side redirect with `getServerProps`?
Of course I can't use both in the same page.
 Maybe it is fine to use some wrapper HOC to the Page component and there define `getInitialProps`? 

thanks!

**Comment by jamesmosier** [comment]
You are able to redirect in `getServerProps` the same as you would in `getInitialProps`:

```js
export async function getServerSideProps(context) {
  if (context.res) {
    // server
    context.res.writeHead(302, {
      Location: '/',
    });

    context.res.end();
  }

  return {
    props: {}, 
  };
}
```

**Reply by oran1248** [reply]
Oh i see... and how it will look in client side transition?

**Reply by jamesmosier** [reply]
It happens on the server so the client will know nothing about the redirect.

**Reply by oran1248** [reply]
So it's not good, because now `getInitialProps` redirects the user

**Reply by jamesmosier** [reply]
I'm not understanding. Could you provide a code sample that outlines what you are asking?

**Reply by oran1248** [reply]
Of course, now i have this code:
```
UserPage.getInitialProps = async (ctx) => {
    if(...user is not authenticated...){
       redirectToMainPage();
    }
}
```

this works for both client transition and server (when user enters page's url in the address bar).
As i see it, with `getServerSideProps` i can't achieve this behavior.

**Reply by Vadorequest** [reply]
Maybe, you can.

getServerSideProps is called from the server only, but during client side transition it's called too (from the server, through an internal invisible ajax call).

So, you may be able to do something similar even for client side redirections.

Alternatively, you can handle that in the render method.

**Reply by oran1248** [reply]
If i'm redirecting in `useEffect`, it will send the request for `getServerSideProps` although it will get "Access denied" or something like that. I'm trying to figure out what is the best way, because it's not clear from the documentation. And the case i'm talking about here is very common...

**Comment by devknoll** [comment]
IMO you should consider something like:

```
function UserPage({ authenticated }) {
  React.useEffect(() => {
    if (!authenticated) {
      Router.push('/login');
    }
  }, [authenticated]);
}

UserPage.getServerSideProps = async (ctx) => {
    if(...user is not authenticated...){
       return { props: { authenticated: false } };
    }
    return { props: { authenticated: true, /* ... */ } };
}
```

**Reply by Vadorequest** [reply]
`getInitialProps` is not deprecated and has valid use-cases, yours may be one.

You cannot redirect from the server when the request comes from the client through Ajax, that's why @devknoll returned `{ authenticated: false }` so that the redirection may happen on the client side.

**Reply by oran1248** [reply]
I didn't say it's deprecated. But I'm talking about a common use case here, so I guess there is another way to achieve that. About @devknoll code,  sorry i thought he used getinitialprops

**Reply by devknoll** [reply]
(I did have `getInitialProps` initially, but it was a typo 😅 . Sorry about that.)

**Reply by oran1248** [reply]
Now I have some HOC that wraps my Page component, for example: withRedux, withToken etc.. can I leave them and use getServerSideProps on the Page component?

**Comment by timneutkens** [comment]
Opened a RFC for redirecting here: https://github.com/vercel/next.js/discussions/14890
